Results Output Generator

lukel43 opened the pull request:
<!-- Replace the title below -->
# Results Output Generator
## Description

This PR adds the the implementation for producing the proper output once checks have been run. This implementation takes the list of check results and sorts and formats it for output to the user. We have also included tests for our implementation in this PR. 

Please keep in mind that we have communicated with other teams to determine the format for the `results` list that we will receive, but it is not yet fully implemented. The code in this PR utilizes a sample `result` that is hard coded, but is the same format we will be receiving during the final implementation. Thanks!
<!-- Remember to check the reminders! -->

## Linked Issues

<!-- Fill in which issue number this PR closes, if there is none, just remove this section! -->
closes: #16 

## Type of Change

<!-- TODO: Fill in the brackets with an `x` next to all types that apply to the proposed changes -->
- [x] Feature
- [ ] Bug fix
- [ ] Documentation

## Contributors

<!-- TODO: Add your GitHub username below and the GitHub usernames of all other contributors to the proposed changes -->
- @lukel43 
- @pkelly55 

## Reminder

 - All GitHub Actions should be in a passing state before any pull request is merged.
 - All PRs must be reviewed by at least one team member and one member of the Integration team!
 - Any issues this PR closes are tagged in the description!

burgess01 commented on gatorgrade/output/pyproject.toml:
```diff
@@ -0,0 +1,12 @@
+name = "poetry-demo"
+version = "0.1.0"
+description = ""
+authors = ["Me lol"]
```
Can you put the authors in here?

lukel43 replied:
This poetry installation was mistakenly included, we were using it to get work done before poetry was installed in the main branch. The whole installation has been removed now!

burgess01 commented on gatorgrade/output/src/output_functions.py:
```diff
@@ -0,0 +1,57 @@
+"""
+This class is used for storing the main functions requested from the Github.
+
+Issue Tracker for the output team.
```
Is this and the previous line of the docstring supposed to be the same sentence? Either way, can you fix the way this sentence is written?

lukel43 replied:
We updated our docstrings to be the correct format and more descriptive, thanks!

burgess01 commented on gatorgrade/output/src/output_functions.py:
```diff
@@ -0,0 +1,57 @@
+    """
+    Process results and determine if the check passed or failed.
+
+    Argument `results` is: list[(check_result_element1, check_result_element2),(...)]]
```
If you are going to use this docstring format, can you use the way we wrote it in the previous practical? ex. Args results: explanation

burgess01 commented on gatorgrade/output/tests/test_function.py:
```diff
@@ -0,0 +1,60 @@
+
+
+def test_false_result_returns_x(capsys):
+    """Test for the X will appear in the output for the failed check."""
```
Can you rephrase this sentence? It reads a little clunky.

burgess01 commented on gatorgrade/output/tests/test_function.py:
```diff
@@ -0,0 +1,60 @@
+    )
+    out, err = capsys.readouterr()
+
+    assert "\u2718" in out
```
In this test you asserted against out, while in the previous one you reassigned out to 'actual_output'. Would you mind making this a uniform choice?

pkelly55 replied:
I made the change to assert actual output for all, thank you.

burgess01 reviewed (changes requested):
Overall this looks pretty good, I just have some corrections for you to make before you merge! Also, I had one more question: I noticed that inside of the output folder you have a tests folder. I am not sure that this follows the file structure we agreed on earlier in the project. We talked about giving each team their own folder, but I'm not sure if this extends to their own test folder as well!

Michionlion commented on gatorgrade/main.py:
```diff
@@ -2,6 +2,7 @@
 from pathlib import Path
 import typer
 
+
```
You shouldn't need to modify this file -- it passes `black` in `main`, and it's best to keep your changes concentrated.
```suggestion
```

lukel43 replied:
Good find! I'm not sure how that happened but I fixed it in my latest push. Thanks!

Michionlion commented on gatorgrade/output/pyproject.toml:
```diff
@@ -0,0 +1,12 @@
+[tool.poetry]
```
It seems that perhaps `poetry add` or some similar command was run in `gatorgrade/output` -- this entire file structure seems pretty wonky. I think you need to take a step back and review where these changes should land. At minimum, there should not be a poetry installation in your `output` folder.

lukel43 replied:
Yes, we had it there before poetry was in the main branch and forgot to uninstall it. This has been fixed in my latest push!

Michionlion commented on gatorgrade/output/src/output_functions.py:
```diff
@@ -0,0 +1,57 @@
+
+# Display a sample output of how the function could display a result object from GatorGrader
+sample_result = [
+    ("No TODOS in text", True),
```
You'll always get a tuple back from GatorGrader that looks like `(description, passed, diagnostic)` -- when `passed` is `True`, `diagnostic` will be `""`.

lukel43 replied:
Thanks for the insight! I have changed the code to include 3 elements for every tuple.

Michionlion commented on gatorgrade/output/tests/test_function.py:
```diff
@@ -0,0 +1,66 @@
+""" Tests to ensure the output_functions.py functions work properly. """
+
+import pytest
+from colorama import init, Fore
```
Instead of importing the individual functions/classes that you're planning on using (especially `init`), it would be better to just `import colorama` (if you'd like a shorter name, you can use `import colorama as color` or something similar) and refer to `init` as `colorama.init()` to avoid confusion.

lukel43 replied:
Great idea, I changed the import to `import colorama as color` as you suggested, which made the `init` statement `color.init()` which is much more readable. Thanks!

Michionlion reviewed (changes requested):
I haven't fully reviewed your work, but there are some changes you'll need to make regarding your file structure, so please make those and reach out if you need details, and then re-request review.

DlHolland reviewed (commented):
Looks good!  The only thing I see is the file structure of output vs src probably needs to be worked out.  

You may want to rename src to something like `display` and put it inside the output folder and make another folder inside `output` that contains the files for running input team's dictionary through gatorgrader, it'll make things a little more organized.

mariakimheinert commented:
Hi @lukel43! Before others continue to review this PR, would you please reorganize the files so that all source code files are in the `gatorgrade/output` directory and all test files are in the top-level `tests/` directory? There should not be any test files in any of the subdirectories of the top-level `gatorgrade/` directory--they should all be in the top-level `tests/` directory. Additionally, there is probably no need to have a separate `src` directory within `gatorgrade/output`; in other words, all source code files should be in the `gatorgrade/output` directory and not in a separate folder called `gatorgrade/output/src`. 

In its current state, the file structure in this PR is difficult to understand and makes reviewing this PR challenging.

lukel43 commented:
Thank you @mariakimheinert, @Michionlion, @burgess01, and @DlHolland for the feedback. I have updated the file structure, and fixed the Poetry situation and Colorama import. I believe I have addressed all requested changes, but during the process our tests started failing. I will work with @pkelly55 to fix the tests so that we can get further feedback. Let me know if there is anything that I failed to address, thanks!

gvanzin-allegheny commented on gatorgrade/output/output_functions.py:
```diff
@@ -1,7 +1,57 @@
-"""Add docstring."""
+"""
+This class is used for storing the main functions requested from the Github.
```
This docstring should match what is located in the command-receive-function branch so that we don't have a merge conflict when combining the two.

The docstring in that branch is:

```
"""This module is used for storing the main functions requested.

The requested functions are located at the Github Issue Tracker
for the output team. For instance, functions dealing with percentage
output, description output, and colorization of text.
"""
```

lukel43 replied:
I decided it was best to use the docstring @gvanzin-allegheny provided given that it will be more consistent, and I feel it is more descriptive than what I had.

gvanzin-allegheny commented on gatorgrade/output/output_functions.py:
```diff
@@ -1,7 +1,57 @@
+    passed_checks = []
+    failed_checks = []
+    # iterate through results tuples
+    for i in results:
```
I would rename the `i` and `j` variables in the loops here to a more descriptive name for better readability.  `i` could be changed to `result`.  Also, rather than looping through `i`, you could instead do : 

```
for i in results:
    pass_check = i[1]
    if pass_check is True:
        passed_checks.append(i)
    else:
        failed_checks.append(i)
```

Since the second element in the tuple will always be a boolean.

lukel43 replied:
I added the proposed changes, thanks!

gvanzin-allegheny commented on gatorgrade/output/output_functions.py:
```diff
@@ -1,7 +1,57 @@
+    """
+    Process results and determine if the check passed or failed.
+
+    Argument `results` is: list[(check_result_element1, check_result_element2),(...)]]
```
There should be a couple extra parentheses here for the argument example that were forgotten:

Argument `results` is: list[(check_result_element1), (check_result_element2),(...)]]

Michionlion replied:
@gvanzin-allegheny @mccurdyc Unfortunately, this data structure (which is a tuple that looks like this: `(description: str, passed: bool, diagnostic: str)`) is provided by an outside tool (GatorGrader), so adding an object to represent it would add on additional translation overhead. This may be the perfect use case for a [namedtuple](https://docs.python.org/3/library/collections.html#collections.namedtuple), however!

lukel43 replied:
I wrote it the way I had it because I was trying to show how there are multiple elements in each check tuple with `(check_result_element1, check_result_element2)`, and that there are multiple checks in the result list by including the `,(...)` portion. However perhaps it could be rewritten to `list[(check_result_element1, check_result_element2, check_result_element3),(...)]`. What do you think?

mccurdyc replied:
@lukel43 I think what you have in the comment is great! Could we use named tuples like @Michionlion suggested? I think that would help a ton from a readability perspective!

Not a huge deal if we can't use named tuples, just a strong suggestion! 😄

lukel43 replied:
Thanks, I will definitely update the comment, but I think it's best to not change the data structure at this time. A more descriptive comment, combined with the other clarity and readability improvements that were made should make the flow of data clear!

gvanzin-allegheny commented on gatorgrade/output/output_functions.py:
```diff
@@ -1,7 +1,57 @@
+
+def output_passed_checks(passed_checks):
+    """Output the results for all of the checks that passed using the passed_checks list."""
+    for i in passed_checks:
```
I would recommend renaming `i` to a more descriptive variable name.

lukel43 replied:
Great idea, I changed it to `check` in my latest push!

gvanzin-allegheny commented on gatorgrade/output/output_functions.py:
```diff
@@ -1,7 +1,57 @@
+
+def output_failed_checks(failed_checks):
+    """Output the results for all of the checks that did not pass using the failed_checks list."""
+    for i in failed_checks:
```
Same issue as above, would rename `i` to a more descriptive name.

lukel43 replied:
Same thing, changed it to `check` in my last push, I think thats the best combination of descriptiveness and simplicity!

mccurdyc commented on tests/test_output.py:
```diff
@@ -0,0 +1,68 @@
+
+def test_output_shows_yellow(capsys):
+    """Testing the failed check will show a yellow message in output."""
+    output_functions.output_failed_checks(
+        failed_checks=("Remove all TODOs", False, "3 TODOs found in example.py")
+    )
```
Why do we expect this to product yellow output for the same input that we provided above that produces red output?

pkelly55 replied:
This test is to prove the diagnostic will be produced for failed checks and the yellow color is properly displayed.

mccurdyc commented on tests/test_output.py:
```diff
@@ -0,0 +1,68 @@
+        failed_checks=("Implement this with an if.", False, "No if statements found")
+    )
+    out, err = capsys.readouterr()
+    expected_output = "\u2718"
```
Is `\u2718` the unicode character for 'x'? Could we define a constant named `unicode_x` or something to help the reader out a bit?

pkelly55 replied:
yes, I have made the change to make a variable for the Unicode string.

mccurdyc reviewed (commented):
Some general comment(s):

1. We should add the `.DS_Store` file to a gitignore (https://stackoverflow.com/a/63517021)
2. I loved your tests! They helped me as the reader understand intent ❤️ 

This is looking good so far! In my opinion, I'd suggest using Objects or dictionaries over using lists/tuples to help the reader out more.

lukel43 commented on gatorgrade/output/output_functions.py:
```diff
@@ -1,7 +1,57 @@
+    for i in failed_checks:
+        # Extract the details of each check
+        requirement = i[0]
+        description = i[2]
```
I like the goal of clarity! As @Michionlion pointed out above, it would likely add more complexity to change the data types, however I did updated the for loops throughout the program to be much more descriptive. Meaning that in the code you pointed out above it now reads, `requirement = check[0]` and `description = check[2]`. I could also add comments to clarify exactly what information is being stored in those variables if necessary. What do you think?

mccurdyc replied:
I think this change to define variables is GREAT!!! And then you don't even need comments to describe what's stored in `check[0]`, etc. The variable name does the work for you! Great change!

lukel43 commented on gatorgrade/output/output_functions.py:
```diff
@@ -1,7 +1,57 @@
+    """
+    Process results and determine if the check passed or failed.
+
+    Argument `results` is: list[(check_result_element1, check_result_element2),(...)]]
```
I agree that the comment could be clarified further, I was attempting to be as descriptive as possible without including a huge docstring. Could you elaborate further on what you mean by "use the language to enforce it"? Thanks!

mccurdyc replied:
@lukel43 great question! Yeah all that I meant by "use the language to enforce it" was could we use an object that had typed fields. Or as @Michionlion has suggested above, I think a named tuple where the tuple elements have types is great!

burgess01 reviewed (approved):
This looks great, good job!

gvanzin-allegheny reviewed (approved):
I think the changes that were made are good.  Nice job

Michionlion commented on gatorgrade/output/output_functions.py:
```diff
@@ -1,7 +1,58 @@
+        return bool
+
+
+# Display a sample output of how the function could display a result object from GatorGrader
+sample_result = [
+    ("No TODOS in text", True, ""),
+    ("Has an if statement", False, "No if statements found"),
+]
+sort_checks_by_result(sample_result)
```
@lukel43 this code is at the top-level of this file and will therefore be run whenever it is imported; this will cause problems for other teams (and your own) as they import your code. Unfortunately, since you merged this without fixing this issue, you'll need to make a hotfix for this so that `main` doesn't break in the future due to this issue.

Michionlion reviewed (commented):
@lukel43 See this critical issue.
